streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -0,0 +1,224 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.internals.CacheFlushListener;
+import org.apache.kafka.streams.kstream.internals.Change;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CachingKeyValueStore<K, V> implements KeyValueStore<K, V> {
+
+    private final KeyValueStore<Bytes, byte[]> underlying;
+    private final Serde<K> keySerde;
+    private final Serde<V> valueSerde;
+    private final CacheFlushListener<K, V> flushListener;
+    private String name;
+    private MemoryLRUCacheBytes cache;
+    private InternalProcessorContext context;
+    private StateSerdes<K, V> serdes;
+    private boolean sendOldValues;
+    private Thread streamThread;
+
+    public CachingKeyValueStore(final KeyValueStore<Bytes, byte[]> underlying,
+                                final Serde<K> keySerde,
+                                final Serde<V> valueSerde,
+                                final CacheFlushListener<K, V> flushListener) {
+        this.underlying = underlying;
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.flushListener = flushListener;
+    }
+
+    @Override
+    public String name() {
+        return underlying.name();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        underlying.init(context, root);
+        initInternal(context);
+        // save the stream thread as we only ever want to trigger a flush
+        // when the stream thread is the the current thread.
+        streamThread = Thread.currentThread();
+    }
+
+    @SuppressWarnings("unchecked")
+    void initInternal(final ProcessorContext context) {
+        this.context = (InternalProcessorContext) context;
+        this.serdes = new StateSerdes<>(underlying.name(),
+                                        keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
+                                        valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+
+        this.name = context.taskId() + "-" + underlying.name();
+        this.cache = context.getCache();
+        cache.addDirtyEntryFlushListener(name, new MemoryLRUCacheBytes.DirtyEntryFlushListener() {
```

Comment:
Why "dirty" -- from my understanding, a "dirty" entry is an invalid one, that does not need to get flushed (actually must not get flushed because it is dirty -- flushing a dirty value would flush something corrupted)